### ISSUE TITLE
[Testing/Usage] How do I write Unit and Integration test that depends on docker?

### ISSUE THREAD

**Issue by xxRockOnxx** [spotify/docker-client#976]
The title says it all. We have a working tests, but this time it includes docker inside it. A cheap test now becomes expensive.

Do I just mock it? But there's a lot of mocking going to happen.

**Comment by mattnworb** [comment]
My advice would be to test code that uses `DockerClient` just like you would treat code that uses a `DatabaseClient` or `AwsS3Client` or any other dependency: when you want to write a unit test of your class that uses some dependency, inject mock versions of the dependency so that you can control the behavior of the `FooClient` and assert how your code reacts in different situations.

If you want to write integration tests to verify that your code + the dependency behave together like you expect, you can certainly do that, but then you introduce a burden of needing a docker daemon that the test can talk to, setting up / cleaning up expected state in the daemon, etc. These tests can certainly have value but also require you to do a lot more in your tests.

So to summarize: treat `DockerClient` like any other dependency of the class being tested.
